docs/posts/ibis-duckdb-geospatial/index.qmd

```diff
@@ -0,0 +1,186 @@
+
+# Download and unzip
+url = "https://open.gishub.org/data/duckdb/nyc_data.db.zip"
+zip_path = Path("nyc_data.db.zip")
+db_path = Path("nyc_data.db")
+
+if not zip_path.exists():
+    urlretrieve(url, zip_path)
+
+if not db_path.exists():
+    with ZipFile(zip_path) as zip_file:
+        zip_file.extract('nyc_data.db')
+```
+
+## Let's get started
+
+The beauty about spatial databases is that they not only allow us to store geometries in them, but also compare them.
+
+```{python}
+import ibis
+from ibis import _
+
+ibis.options.interactive = True
+
+con = ibis.duckdb.connect("nyc_data.db")
+con.list_tables()
+```
+
+In this database we have multiple tables with information about New York City. Following, Dr. Qiusheng Wu's class, we'll take a look at some spatial relations.
+
+We can start by taking a peak at the `nyc_subway_stations` table.
+
+```{python}
+subway_stations = con.table('nyc_subway_stations')
+subway_stations
+```
+
+Notice that the last column has a `geometry` type, and in this case it contains a points that represent the location of each subway station. Let's take the entry for `'Broad St'`.
+
+```{python}
+broad_station =  subway_stations.filter(subway_stations.NAME == 'Broad St')
+broad_station
+```
+
+### `geo_equals` (`ST_Equals`)
+
+In DuckDB `ST_Equals` returns `True` if two geometries are topologically equal, meaning that they have the same dimension and they have identical coordinate values, although the order of the vertices may be different.
+
+The following  is a bit redundant but we can check if our broad station point
+matches only one point on our data using `geo_equals`
+
+```{python}
+subway_stations.filter(subway_stations.geom.geo_equals(broad_station.geom))
+```
+
+We can also write this query without using `broad_station` as a variable, and with the help of the deferred expressions API, also known as Underscore (`_`) API (Learn more about it in the [chaining expressions docs](https://ibis-project.org/how-to/analytics/chain_expressions.html))
+
+```{python}
+subway_stations.filter(_.geom.geo_equals(_.filter(_.NAME == 'Broad St').geom))
+```
+
+### `intersect` (`ST_Intersect)
+
+Let's find in which neighborhood is "Broad Street" subway station and determine its neighborhood using the the geospatial `intersect` function. The `intersect` function returns `True` if two geometries have any points in common.
+
+The
+
+```{python}
+boroughs = con.table('nyc_neighborhoods')
+boroughs
+```
+
+
+### FIXME: NEED AN EXPLANATION ABOUT WHY
+it's still unclear to me why:
+a) this works broad_station.select(broad_station.geom).to_array())
+b) but this doesn't broad_station.geom.to_array()).  -> AttributeError: 'GeoSpatialColumn' object has no attribute 'to_array'
+
+Looks like a) is ibis.expr.types.relations.Table while b) ibis.expr.types.geospatial.GeoSpatialColumn
+find a way to explain why we do this
+
+```{python}
+boroughs.filter(boroughs.geom.intersects(broad_station.select(broad_station.geom).to_array()))
```

Comment:
It's still not clear to me how to explain why we have to do this `broad_station.select(broad_station.geom).to_array())` for things to work and this `broad_station.geom.to_array())`  leads to `AttributeError: 'GeoSpatialColumn' object has no attribute 'to_array'` 

Looks like broad_station.select(broad_station.geom) is an `ibis.expr.types.relations.Table` while `broad_station.geom` is  `ibis.expr.types.geospatial.GeoSpatialColumn`

Any help doing this would be appreciated it.

Reply:
@cpcloud would you give me hand explaining the reasoning behind having to do  the `select` here?

Reply:
Definitely.

`to_array()` is an to generate a scalar subquery from table with a single column (whose shape is scalar).

In SQL, the way this is done is by wrapping a `SELECT` statement in `()`:

```sql
SELECT x < (SELECT avg(x) FROM t)
FROM t
```

Which roughly corresponds to this Ibis code:

```python
t.select(_.x < t.agg(_.x.mean()).to_array())
```

`Column`s don't have this method because no one has implemented it, plus `to_array()` is a name from Ibis "Classic" that we should change because it no longer maps to any concepts in Ibis so we shouldn't add it to other things before renaming it to something else like `subq`/`subquery` etc.